### PR TITLE
feat: openapi-implementation

### PR DESCRIPTION
**Pull request by dasarathg68**
<h1> Description </h1>
The PR filed addresses issue #12 and presents OpenAPI specification for the Node.js backend.
swagger-autogen package is used to automatically parse the routes and identify paths, parameters and also descriptions.
The generated document is served using swagger-ui-express npm package on the /api-docs endpoint.
Further, the controller functions are grouped using by using tags. (Teams, Members)
Here's how it looks.
Screenshot 2024-03-30 at 12 07 56 PM
Screenshot 2024-03-30 at 12 08 31 PM

<h1> Type of Change </h1>
Non breaking change that implements OpenAPI for description of the APIs.

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -3,6 +3,11 @@
   "version": "1.0.0",
   "description": "",
   "main": "src/index.ts",
+  "nodemonConfig": {
+    "ignore": [
+      "src/utils/swagger-output.json"
+    ]
+  },
```
Comment (hermannleboss) [inline backend/package.json:10]: Why this is necessary

File: backend/src/controllers/memberController.ts [hunk 2]
```diff
@@ -50,10 +56,13 @@ const updateMember = async (req: Request, res: Response) => {
...
 //delete members
 const deleteMembers = async (req: Request, res: Response) => {
+  /* 
+  #swagger.tags = ['Members']
+  */
   const { id } = req.params;
 
   try {
-    const member = await prisma.member.delete({
+    const member = await prisma.member.deleteMany({
```
Comment (hermannleboss) [inline backend/src/controllers/memberController.ts:65]: Why this change
And even if it's an optimization or a correction, it has nothing to do with swagger and open API.
So it shouldn't be in this PR

File: backend/package.json
Comment (hermannleboss) [inline backend/package.json:1]: Look at this if it can be helpfull
https://www.npmjs.com/package/swagger-jsdoc

File: backend/src/utils/swagger-output.json
Comment (hermannleboss) [inline backend/src/utils/swagger-output.json:1]: If it's generated everytime we run the server do we need to have it inside git ?

**Review by hermannleboss — CHANGES_REQUESTED** [review]
Good starting point. 

Look at this API,
https://petstore.swagger.io/


The api have to explain herself. 

- [x] Every input should have a type, 
- [x] we should know if it's optional or not. 
- [ ] And we should know the type of the output 
- [ ] and his status

Try to implement theses requirements

File: backend/package.json (hunk 1 above)
Comment (dasarathg68) [reply backend/package.json:10]: During dev we use nodemon to watch for changes right? 
It creates an infinite loop as every time the swagger-output is generated, it assumes it to be a change and restarts the server over and over again. 
Thats why this has to be present

File: backend/src/controllers/memberController.ts (hunk 2 above)
Comment (dasarathg68) [reply backend/src/controllers/memberController.ts:65]: This change is about swagger.
swagger assumes anything in the pattern .delete() to be api endpoints.
So prisma.member.delete() would be considered to be an api endpoint which is not the case.

File: backend/package.json
Comment (dasarathg68) [reply backend/package.json:1]: Thanks but I already took a look at that package and it cannot be used since we're not generating openapi specification. We're using swagger-autogen library which uses minimal directives to generate the specification automatically and swagger-ui-express to render the specification.

File: backend/src/utils/swagger-output.json
Comment (dasarathg68) [reply backend/src/utils/swagger-output.json:1]: We need the file to be present. It is a requirement of the swagger-autogen library.
If we don't add to git then we have to create it manually each time.
Atleast an empty file named swagger-output.json with just {} should be present

File: backend/package.json (hunk 1 above)
Comment (hermannleboss) [reply backend/package.json:10]: OK

**Comment by dasarathg68** [comment]
> Good starting point.
> 
> Look at this API, https://petstore.swagger.io/
> 
> The api have to explain herself.
> 
> * [x]  Every input should have a type,
> * [x]  we should know if it's optional or not.
> * [x]  And we should know the type of the output
> * [x]  and his status
> 
> Try to implement theses requirements

@hermannleboss i have updated/reverted the code to work with the manual version and i have addressed the requirements. Please check the documentation at localhost:3000/api-docs after running the backend.

**Review by hermannleboss — APPROVED** [review]
